### PR TITLE
scm/git UI fixes

### LINKED ISSUE #5245: [scm] wrong icon for open changes
before the merge of the SCM PR it looked like this:

Screen Shot 2019-05-24 at 13 00 24

and now:

Screen Shot 2019-05-24 at 13 00 37

@vinokurig, cf. https://github.com/theia-ide/theia/pull/4279#issuecomment-492511793, @akosyakov's comment was not addressed

### PR DESCRIPTION
**Pull request by vinokurig**
- [x] fix https://github.com/theia-ide/theia/issues/5245
- [x] fix https://github.com/theia-ide/theia/issues/5267
- [x] fix https://github.com/theia-ide/theia/issues/3800
- [x] fix https://github.com/theia-ide/theia/issues/5266
- [x] fix https://github.com/theia-ide/theia/issues/5262
- [x] fix https://github.com/theia-ide/theia/issues/5246
- [x] fix https://github.com/theia-ide/theia/issues/5265 (see https://github.com/theia-ide/theia/issues/5265#issuecomment-496215695)
- [x] fix https://github.com/theia-ide/theia/issues/5263
- [x] fix https://github.com/theia-ide/theia/issues/4851
- [x] fix https://github.com/theia-ide/theia/issues/5291

### CODE REVIEW

**Comment by akosyakov** [comment]
Let's keep this PR to recover all lost changes in one commit that we don't create too much noise with fixing what already used to work. I'm going to go through git extension history starting from Jan now and list commits which should be recovered here. Hope is not much.

**Comment by vinokurig** [comment]
@akosyakov OK, merged previous PRs

**Comment by akosyakov** [comment]
@vinokurig ok, please close other PRs then, I am half through changes, will check/test the rest after the lunch. I open issues for everything which i find and will retest them against this PR after analyzing commits.

File: packages/git/src/browser/git-contribution.ts [hunk 1]
```diff
@@ -148,7 +148,8 @@ export namespace GIT_COMMANDS {
     export const OPEN_CHANGES: Command = {
         id: 'git.open.changes',
         category: 'Git',
-        label: 'Open Changes'
+        label: 'Open Changes',
+        iconClass: 'theia-open-change-icon'
```
Comment (akosyakov) [inline packages/git/src/browser/git-contribution.ts]: here is the original changes: https://github.com/theia-ide/theia/commit/2519e5323cc6bdaa2bc4bccaabc6abf83dd81aaf#diff-23f6de2d4f77ec14174dd983bef7b2cc

please apply them, changes to `Open File` command should be recovered as well

**Comment by akosyakov** [comment]
@vinokurig I went through all changes from Jan and updated a comment with referenced issues. I'm starting testing and will make verified as done.

**Comment by akosyakov** [comment]
@vinokurig please try to toggle SCM several times:
togglescm

icon and widget caption are not rendered after toggling couple times

File: packages/scm/src/browser/scm-widget.tsx [hunk 2]
```diff
@@ -115,7 +118,7 @@ export class ScmWidget extends ScmNavigableListWidget<ScmResource> implements St
         this.scmService.onDidChangeSelectedRepositories(repository => {
             if (repository) {
                 this.selectedRepoUri = repository.provider.rootUri;
-                this.title.label = 'Source Control: ' + repository.provider.contextValue;
+                this.title.label = ScmWidget.LABEL + ': ' + repository.provider.contextValue;
```
Comment (akosyakov) [inline packages/scm/src/browser/scm-widget.tsx]: Should not we update title and rerender the widget when `repository` is `undefined` as well?

**Comment by akosyakov** [comment]
@vinokurig https://github.com/theia-ide/theia/pull/5254/files#diff-a2eb5b2bcd08435891fb7969b75c9fccL119 looks bogus, it means if you move git widget to any other area except left updating title won't work anymore.

File: packages/scm/src/browser/scm-widget.tsx [hunk 3]
```diff
@@ -115,7 +118,7 @@ export class ScmWidget extends ScmNavigableListWidget<ScmResource> implements St
         this.scmService.onDidChangeSelectedRepositories(repository => {
             if (repository) {
```
Comment (akosyakov) [inline packages/scm/src/browser/scm-widget.tsx]: when we toggle the widget, old one gets destroyed and new one is created, it looks like for second time this code is never executed

File: packages/scm/src/browser/scm-widget.tsx (hunk 2 above)
Comment (vinokurig) [reply packages/scm/src/browser/scm-widget.tsx]: Fixed

**Comment by vinokurig** [comment]
@akosyakov 
> @vinokurig please try to toggle SCM several times:

Fixed

File: packages/scm/src/browser/scm-widget.tsx (hunk 2 above)
Comment (akosyakov) [reply packages/scm/src/browser/scm-widget.tsx]: > Fixed

@vinokurig  maybe you meant https://github.com/theia-ide/theia/pull/5254/files#r287785296?

Regarding this comment, widget and title don't get updated in `else` case

Comment (vinokurig) [reply packages/scm/src/browser/scm-widget.tsx]: Sorry forgot to push

Comment (akosyakov) [reply packages/scm/src/browser/scm-widget.tsx]: ok, let me know when you push

Comment (vinokurig) [reply packages/scm/src/browser/scm-widget.tsx]: already pushed

**Comment by vinokurig** [comment]
@akosyakov 
> @vinokurig https://github.com/theia-ide/theia/pull/5254/files#diff-a2eb5b2bcd08435891fb7969b75c9fccL119 looks bogus, it means if you move git widget to any other area except left updating title won't work anymore.

Fixed

File: packages/scm/src/browser/scm-widget.tsx (hunk 3 above)
Comment (vinokurig) [reply packages/scm/src/browser/scm-widget.tsx]: I've tested and it works fine

**Comment by akosyakov** [comment]
@vinokurig everything seems to work now, except #3800

File: packages/scm/src/browser/scm-widget.tsx [hunk 4]
```diff
@@ -115,15 +114,30 @@ export class ScmWidget extends ScmNavigableListWidget<ScmResource> implements St
...
                 this.selectedRepoUri = repository.provider.rootUri;
-                this.title.label = 'Source Control: ' + repository.provider.contextValue;
-                this.shell.leftPanelHandler.refresh();
+                this.title.label = ScmWidget.LABEL + ': ' + repository.provider.contextValue;
+                const area = this.shell.getAreaFor(this);
```
Comment (akosyakov) [inline packages/scm/src/browser/scm-widget.tsx]: BTW why do we need to refresh an area? I've tried to remove it and could not break anything

File: packages/scm/src/browser/scm-widget.tsx [hunk 5]
```diff
@@ -115,15 +114,30 @@ export class ScmWidget extends ScmNavigableListWidget<ScmResource> implements St
...
+                if (area === 'left') {
+                    this.shell.leftPanelHandler.refresh();
+                } else if (area === 'right') {
+                    this.shell.rightPanelHandler.refresh();
+                }
                 this.update();
```
Comment (akosyakov) [inline packages/scm/src/browser/scm-widget.tsx]: we should call `update` always

**Comment by akosyakov** [comment]
@vinokurig are you looking into #3800? i can do if you don't have time, we need to resolve and merge it  today

**Comment by vinokurig** [comment]
@akosyakov 
> @vinokurig are you looking into #3800? i can do if you don't have time, we need to resolve and merge it today

yes, working on it right now

**Comment by vinokurig** [comment]
@akosyakov #3800 is fixed could you please check

File: packages/scm/src/browser/scm-widget.tsx (hunk 4 above)
Comment (vinokurig) [reply packages/scm/src/browser/scm-widget.tsx]: It is used to update the title of the tab

File: packages/scm/src/browser/scm-widget.tsx (hunk 5 above)
Comment (vinokurig) [reply packages/scm/src/browser/scm-widget.tsx]: fixed

**Comment by vinokurig** [comment]
@akosyakov looks like I've answered all the questions, please point me if I missed something

File: packages/git/src/browser/git-repository-provider.ts [hunk 6]
```diff
@@ -93,6 +94,9 @@ export class GitRepositoryProvider {
      * Sets or un-sets the repository.
      */
     set selectedRepository(repository: Repository | undefined) {
+        if (!repository) {
+            this.scmService.selectedRepository = repository;
+        }
         this._selectedRepository = repository;
```
Comment (akosyakov) [inline packages/git/src/browser/git-repository-provider.ts]: - What is the difference between `selectedRepository` in `GitRepositoryProvider` and `ScmService`?
- Should not they point to the same repository? Can `GitRepositoryProvider` to `ScmService` for `selectedRepository` and `onDidChangeRepository` event?

File: packages/git/src/browser/git-repository-provider.ts [hunk 7]
```diff
@@ -93,6 +94,9 @@ export class GitRepositoryProvider {
      * Sets or un-sets the repository.
      */
     set selectedRepository(repository: Repository | undefined) {
+        if (!repository) {
```
Comment (akosyakov) [inline packages/git/src/browser/git-repository-provider.ts]: Why `undefined` should be ignored?

**Comment by akosyakov** [comment]
@vinokurig something wrong with git state still, after deleting the repo not all status bar elements are removed:

Screen Shot 2019-05-28 at 10 30 17

Sync changes should not be there.

**Comment by akosyakov** [comment]
@vinokurig it looks really bad that there is no a single source of a currently selected repository for frontend, git and scm seems to manage their own and try to sync them. It should be refactored to have this state only once and remove all sync logic. Do you think it is possible to accomplish today?

**Comment by vinokurig** [comment]
@akosyakov 
> @vinokurig it looks really bad that there is no a single source of a currently selected repository for frontend, git and scm seems to manage their own and try to sync them. It should be refactored to have this state only once and remove all sync logic. Do you think it is possible to accomplish today?

I'll try to do it today

**Comment by vinokurig** [comment]
@akosyakov 
There is `GitRepositoryProvider` that provides Git specific repositories. The selected repository in `GitRepositoryProvider` is set by SCM service. To get rid of `GitRepositoryProvider` it is needed to completely rework Git extension and other related extensions. It looks like very big part of work. 
I agree that firing repository deleted event from Git extension is not good, but I can't detect deletion of projects folder in the Scm because File watcher event returns only an array of deleted files.

**Comment by akosyakov** [comment]
I don't mean to remove `GitRepositoryProvider` but make it git specific facade of `ScmService`. Handling of `ScmProvider` also should be moved to `GitRepositoryProvider` instead of keeping `allRepositories`.

`ScmService` should be a single source of the whole scm state, git extension should populate it and provide git specific instead of duplicating this state partially and trying to sync it.

Ok, let's open an issue to refactor it.

**Comment by akosyakov** [comment]
Have you managed to track down https://github.com/theia-ide/theia/pull/5254#issuecomment-496418791?

Why status bar elements are not removed.

**Comment by vinokurig** [comment]
> Have you managed to track down #5254 (comment)?

Why status bar elements are not removed.

working on it now

**Comment by akosyakov** [comment]
Here https://github.com/theia-ide/theia/blob/59fdb74937de03764ec9cdb6255dbcd7832c55a2/packages/git/src/browser/git-contribution.ts#L274-L288

I wonder whether we should reset to `undefined` selected repository for git extension as well. Otherwise it keeps a reference to a repository which is not selected anymore. Maybe it is a reason.

**Comment by vinokurig** [comment]
@akosyakov Fixed icon cleanup

**Comment by vinokurig** [comment]
@akosyakov any updates?

**Comment by akosyakov** [comment]
now it cannot detect when a new repository is cloned after removing

Screen Shot 2019-05-28 at 16 16 47

Strangely it is detected in the status bar, but not in SCM view. Order of items also is bogus.

Also if you have a folder opened with 2 repos and delete one. The status bar item to pick a repo should disappear and appear when this repo is cloned again. It was working nicely before.

**Comment by vinokurig** [comment]
@akosyakov Is it ok to fix it tomorrow?

**Comment by akosyakov** [comment]
@vinokurig after testing more previous state: it is fine that new repositories cannot be detected without explicit refresh, but then status bar should not snow bogus elements like in https://github.com/theia-ide/theia/pull/5254#issuecomment-496537387

With explicit refresh with multiple git repositories, all new repositories should be detected.

**Comment by vinokurig** [comment]
@akosyakov Fixed icons order

**Comment by akosyakov** [comment]
@vinokurig ok, testing, are you looking into detection addition/deletion repositories on refresh?

**Comment by vinokurig** [comment]
> @vinokurig ok, testing, are you looking into detection addition/deletion repositories on refresh?

yes

**Comment by akosyakov** [comment]
When I delete the repo it still shows a branch picker in the status bar:
Screen Shot 2019-05-29 at 09 28 21

and when i clone it again, it shows SCM elements in status bar, although it should show anything (i did not trigger refresh):
Screen Shot 2019-05-29 at 09 29 29

It seems that somewhere repository is cached. For right now i see state is duplicated several times:
- `GitContribution.dirtyResporitories` duplicating `GitRepositoryProvider.allRepositories` which duplicating `ScmService.repositories`
- and what discussed yesterday `GitRepositoryProvider.selectedRepository` duplicating `ScmService.selectedRepository`

I suspect deleted repository is still cached somewhere.

**Comment by vinokurig** [comment]
@akosyakov detection addition/deletion repositories is fixed

**Comment by vinokurig** [comment]
@akosyakov
> when i clone it again, it shows SCM elements in status bar, although it should show anything (i did not trigger refresh

couldn't reproduce, maybe it is fixed with the latest commits

**Comment by akosyakov** [comment]
@vinokurig thank you for all effort! it works good now. Please squash commits, we are waiting for travis to be green then.

**Comment by marcdumais-work** [comment]
@akosyakov So if I understand correctly, this is not blocking the `0.7.0` release any more?

**Comment by akosyakov** [comment]
@marcdumais-work yes, it should be good for 0.7.0

**Comment by akosyakov** [comment]
@vinokurig Could you have a look at https://github.com/theia-ide/theia/issues/5300? i've missed during testing :(

**Comment by vinokurig** [comment]
@akosyakov yes, working on it

**Comment by vinokurig** [comment]
@akosyakov 
> @vinokurig Could you have a look at #5300? i've missed during testing :(

fixed

**Pull request merged**